File: Outco/heapSort.js
/*
 *  Target Practice 08 - Heapsort
 */

'use strict';

/*
 *  Heapsort
 *
 *  Prompt:    Given an unsorted array of integers, return the array
 *             sorted using heapsort.
 *
 *  Input:     input {Array}
 *  Output:    {Array}
 *
 *  Example:   [3,9,1,4,7] --> [1,3,4,7,9]
 */

// Worse Time Complexity:
// Worse Auxiliary Space Complexity:
// Average Time Complexity:
// Average Auxiliary Space Complexity:


function heapsort(arr) {

  // Convert unsorted array to maxheap

  // console.log(arr);
  for (let i = arr.length - 1; i > -1; i--) {
    bubbleDown(i, arr, arr.length);
  }

  // console.log(arr);



  // Convert maxheap to sorted array
  let boundary = arr.length;

  while (boundary > 0) {
    swap(arr, 0, boundary - 1);
    boundary--;
    bubbleDown(0, arr, boundary);
  }

  // console.log(arr);
  return arr;
}

function bubbleDown(parentIndex, arr, boundary) {
  let childIndex = getChildIndex(parentIndex, arr, boundary);

  while (childIndex < boundary && arr[parentIndex] < arr[childIndex]) {
    swap(arr, parentIndex, childIndex);
    parentIndex = childIndex;
    childIndex = getChildIndex(parentIndex, arr, boundary);
  }
};

function getChildIndex(parentIndex, arr, boundary) {
  let childIndex1 = 2 * parentIndex + 1;
  let childIndex2 = 2 * parentIndex + 2;

  if (childIndex1 >= boundary) {
    return childIndex1;
  } else if (childIndex2 >= boundary) {
    return childIndex1;
  } else if (arr[childIndex1] > arr[childIndex2]) {
    return childIndex1;
  } else {
    return childIndex2;
  }

}

function swap(arr, i, j) {
  [arr[i], arr[j]] = [arr[j], arr[i]];
}




////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////


// custom assert function to handle tests
// input: count {Array} - keeps track out how many tests pass and how many total
//        in the form of a two item array i.e., [0, 0]
// input: name {String} - describes the test
// input: test {Function} - performs a set of operations and returns a boolean
//        indicating if test passed
// output: {undefined}
function assert(count, name, test) {
  if(!count || !Array.isArray(count) || count.length !== 2) {
    count = [0, '*'];
  } else {
    count[1]++;
  }

  let pass = 'false';
  let errMsg = null;
  try {
    if (test()) {
      pass = ' true';
      count[0]++;
    }
  } catch(e) {
    errMsg = e;
  }
  console.log('  ' + (count[1] + ')   ').slice(0,5) + pass + ' : ' + name);
  if (errMsg !== null) {
    console.log('       ' + errMsg + '\n');
  }
}


// check if an array is sorted in linear time
function isSorted(arr) {
  if (arr.length < 2){ return true; }
  for (let i = 1; i < arr.length; i++){
    if (arr[i-1] > arr[i]) { return false; }
  }
  return true;
}


// compare if two flat arrays are equal
function arraysEqual(arr1, arr2) {
  if (arr1.length !== arr2.length) { return false; }
  for (let i = 0; i < arr1.length; i++) {
    if (arr1[i] !== arr2[i]) { return false; }
  }
  return true;
}

console.log('heapsort tests');
let testCount = [0, 0];

assert(testCount, 'able to sort a single-element array', () => {
  let results = heapsort([5]);
  return arraysEqual(results, [5]);
});

assert(testCount, 'able to sort a medium-sized unsorted array', () => {
  let results = heapsort([4,15,16,50,8,23,42,108]);
  return arraysEqual(results, [4,8,15,16,23,42,50,108]);
});

assert(testCount, 'able to sort a large unsorted array', () => {
  let test = [];
  for (let i = 0; i < 1000000; i++){
    test.push(Math.floor(Math.random() * 1000000));
  }
  let solution = heapsort(test.slice());
  test = test.sort((a , b) => { return a - b; });
  return arraysEqual(solution, test);
});
console.log('PASSED: ' + testCount[0] + ' / ' + testCount[1], '\n\n');

File: Outco/heapSort.py
#  Target Practice 08 - HeapSort
#
#  Prompt: Implement heapsort in-place.
#
#  NOTE: In-place means to manipulate the input list rather than create a
#        new list.
#
#  Input: {List}
#  Output: {List}
#
#  Example: heapsort([4,15,16,50,8,23,42,108])
#           // [4,8,15,16,23,42,50,108]
#

# Worse Time Complexity:
# Worse Auxiliary Space Complexity:
# Average Time Complexity:
# Average Auxiliary Space Complexity:

def build_heap(lst):
  first_parent_idx = (len(lst) - 2)//2
  for i in reversed(range(first_parent_idx + 1)):
    bubble_down(i, len(lst) - 1, lst)

def bubble_down(curr_idx, end_idx, lst):
  child_1_idx = 2*curr_idx + 1
  while child_1_idx <= end_idx:
    temp = 2*curr_idx + 2
    child_2_idx = temp if temp <= end_idx else -1
    if child_2_idx > -1 and lst[child_2_idx] > lst[child_1_idx]:
      child_to_swap = child_2_idx
    else:
      child_to_swap = child_1_idx
    if lst[child_to_swap] > lst[curr_idx]:
      lst[child_to_swap], lst[curr_idx] = lst[curr_idx], lst[child_to_swap]
      curr_idx = child_to_swap
      child_1_idx = 2*curr_idx + 1
    else:
      return 

def heapsort(lst):
  build_heap(lst)
  for i in reversed(range(1, len(lst))):
    lst[i], lst[0] = lst[0], lst[i]
    bubble_down(0, i - 1, lst)
  return lst

############################################################
###############  DO NOT TOUCH TEST BELOW!!!  ###############
############################################################

# custom expect function to handle tests
# List count : keeps track out how many tests pass and how many total
#   in the form of a two item array i.e., [0, 0]
# String name : describes the test
# Function test : performs a set of operations and returns a boolean
#   indicating if test passed
def expect(count, name, test):
    if (count is None or not isinstance(count, list) or len(count) != 2):
        count = [0, 0]
    else:
        count[1] += 1

    result = 'false'
    error_msg = None
    try:
        if test():
            result = ' true'
            count[0] += 1
    except Exception as err:
        error_msg = str(err)

    print('  ' + (str(count[1]) + ')   ') + result + ' : ' + name)
    if error_msg is not None:
        print('       ' + error_msg + '\n')

# code for capturing print output


from io import StringIO
import sys


class Capturing(list):
    def __enter__(self):
        self._stdout = sys.stdout
        sys.stdout = self._stringio = StringIO()
        return self

    def __exit__(self, *args):
        self.extend(self._stringio.getvalue().splitlines())
        sys.stdout = self._stdout

# code for checking if lists are equal
def lists_equal(lst1, lst2):
    if len(lst1) != len(lst2):
        return False
    for i in range(0, len(lst1)):
        if lst1[i] != lst2[i]:
            return False
    return True

# custom function for checking if list is sorted (linear runtime)
def is_sorted(input):
    if (len(input) < 2):
        return True
    for i in range(1, len(input)):
        if (input[i-1] > input[i]):
            return False
    return True

# for getting random numbers
from random import randint

print('heapsort tests')
test_count = [0, 0]


def test():
    results = heapsort([5])
    return is_sorted(results)


expect(test_count, 'able to sort a single-element array', test)


def test():
    results = heapsort([4, 15, 16, 50, 8, 23, 42, 108])
    return lists_equal(results, [4, 8, 15, 16, 23, 42, 50, 108])


expect(test_count, 'able to sort a medium-sized unsorted list', test)


def test():
    results = []
    for i in range(0, 1000000):
        results.append(int(randint(0, 1000000)))
    example = heapsort(results)
    return lists_equal(sorted(results), example)


expect(test_count, 'able to sort a large unsorted array', test)
